storage/storage.rs

```
@@ -550,56 +356,137 @@ impl<D> MVCCStorage<D> {
+) -> Result<(IsolationManager, Vec<SequenceNumber>), StorageCommitError> {
+    use StorageCommitError::{Durability, IO};
+
+    let isolation_manager = IsolationManager::new(next_sequence_number);
+
+    enum CheckpointCommitStatus {
+        Pending(CommitRecord),
+        Validated(Option<CommitRecord>),
+        Rejected,
+    }
+
+    let mut recovered_commits = BTreeMap::new();
+    for record in durability_service
+        .iter_from(next_sequence_number)
+        .map_err(|error| Durability { name: storage_name.to_owned(), source: error })?
+    {
+        let RawRecord { sequence_number, record_type, bytes } =
+            record.map_err(|error| IO { name: storage_name.to_owned(), source: error })?;
+        match record_type {
+            CommitRecord::RECORD_TYPE => {
+                let commit_record = CommitRecord::deserialise_from(&mut &*bytes).unwrap(); // FIXME
+                recovered_commits.insert(sequence_number, CheckpointCommitStatus::Pending(commit_record));
+            }
+            StatusRecord::RECORD_TYPE => {
+                let StatusRecord { commit_record_sequence_number, was_committed } =
+                    StatusRecord::deserialise_from(&mut &*bytes).unwrap(); // FIXME
+                if was_committed {
+                    let record = recovered_commits.remove(&commit_record_sequence_number).map(|status| {
+                        let CheckpointCommitStatus::Pending(record) = status else {
+                            unreachable!("found second commit status for a record")
+                        };
+                        record
+                    });
+                    recovered_commits.insert(commit_record_sequence_number, CheckpointCommitStatus::Validated(record));
+                } else {
+                    recovered_commits.insert(commit_record_sequence_number, CheckpointCommitStatus::Rejected);
+                }
+            }
+            _other => (), // skip?
```

This should be a unreachable - we should stay away from code branches that will silently continue and let us get away with not updating all affected branches of code if something changes (like a new record type).